notifier/core/params.py

```diff
@@ -0,0 +1,135 @@
+import os
+import sys
+import json
+from .cli import CLI
+from .config import Config
+
+
+class Params:
+    def __init__(self):
+        self.__default = {
+            "plugin_path": "/etc/notifier/plugins",
+            "broker_url": "",
+            "queue_name": "notifier_queue",
+            "hostname": "notifier",
+            "num_workers": 1,
+            "log_from": "INFO",
+            "log_path": "/var/logs",
+            "pid_path": "/var/run"
+        }
+        self.__cli = CLI()
+        self.__args, unk = self.__cli.parse_known_args()
+        self.__check_default()
+        self.__config = Config()
+
+    def __check_default(self):
+        if self.__args.default_params:
+            self.__print_default()
+            sys.exit(0)
+
+    def __print_default(self):
+        print(json.dumps(self.__default, indent=4))
+
+    def plugin_path(self):
+        if self.__args.plugin_path:
+            return self.__args.plugin_path
+
+        plugin_path_os = os.environ.get("NOTIFIER_PLUGIN_PATH")
+        if plugin_path_os:
+            return plugin_path_os
+
+        if self.__config.key("plugin_path"):
+            return self.__config.key("plugin_path")
+
+        return self.__default["plugin_path"]
+
+    def broker_url(self):
+        if self.__args.broker_url:
+            return self.__args.broker_url
+
+        broker_url_os = os.environ.get("NOTIFIER_BROKER_URL")
+        if broker_url_os:
+            return broker_url_os
+
+        if self.__config.key("broker_url"):
+            return self.__config.key("broker_url")
+
+        return None
+
+    def queue_name(self):
+        if self.__args.queue_name:
+            return self.__args.queue_name
+
+        queue_name_os = os.environ.get("NOTIFIER_QUEUE_NAME")
+        if queue_name_os:
+            return queue_name_os
+
+        if self.__config.key("queue_name"):
+            return self.__config.key("queue_name")
+
+        return self.__default["queue_name"]
+
+    def hostname(self):
+        if self.__args.hostname:
+            return self.__args.hostname
+
+        hostname_os = os.environ.get("NOTIFIER_HOSTNAME")
+        if hostname_os:
+            return hostname_os
+
+        if self.__config.key("hostname"):
+            return self.__config.key("hostname")
+
+        return self.__default["hostname"]
+
+    def num_workers(self):
+        if self.__args.num_workers:
+            return self.__args.num_workers
+
+        num_workers_os = os.environ.get("NOTIFIER_NUM_WORKERS")
+        if num_workers_os:
+            return num_workers_os
+
+        if self.__config.key("num_workers"):
+            return self.__config.key("num_workers")
+
+        return self.__default["num_workers"]
+
+    def log_from(self):
```

Comment:
This implementation doesn't work. For some reason, the workers read from args and return the default value `/var/logs/worker1.log`

This is was tested in the following way:
1. Add some prints in each return 
2. Launch notifier

The prints in this functions are the following
```python
    def log_from(self):
        if self.__args.log_from:
            print("--- 1")
            return self.__args.log_from

        log_from_os = os.environ.get("NOTIFIER_LOG_FROM")
        if log_from_os:
            print("--- 2")
            return log_from_os

        if self.__config.key("log_from"):
            print("--- 3")
            return self.__config.key("log_from")

        print("--- 4")
        return self.__default["log_from"]
```
The result after of install the app with pip, define the `NOTIFIER_CONFIG` env var and launch using `python -m notifier`  the result is the following 

```
$ python -m notifier
--- 3
--- 3
celery multi v4.1.0 (latentcall)
> Starting nodes...
--- 1
	> worker1@notifier: OK
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/__main__.py", line 18, in <module>
    main()
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/__main__.py", line 14, in main
    _main()
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 326, in main
    cmd.execute_from_commandline(argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 488, in execute_from_commandline
    super(CeleryCommand, self).execute_from_commandline(argv)))
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/base.py", line 281, in execute_from_commandline
    return self.handle_argv(self.prog_name, argv[1:])
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 480, in handle_argv
    return self.execute(command, argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 412, in execute
    ).run_from_argv(self.prog_name, argv[1:], command=argv[0])
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/worker.py", line 220, in run_from_argv
    self.maybe_detach([command] + argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/worker.py", line 227, in maybe_detach
    detached_celeryd(self.app).execute_from_commandline(argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celeryd_detach.py", line 96, in execute_from_commandline
    **vars(options)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celeryd_detach.py", line 32, in detach
    after_forkers=False):
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/platforms.py", line 334, in open
    self.after_chdir()
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/platforms.py", line 411, in after_chdir_do
    logfile and open(logfile, 'a').close()
FileNotFoundError: [Errno 2] No such file or directory: '/var/logs/worker1.log'
--- 1
	> worker2@notifier: OK
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/__main__.py", line 18, in <module>
    main()
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/__main__.py", line 14, in main
    _main()
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 326, in main
    cmd.execute_from_commandline(argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 488, in execute_from_commandline
    super(CeleryCommand, self).execute_from_commandline(argv)))
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/base.py", line 281, in execute_from_commandline
    return self.handle_argv(self.prog_name, argv[1:])
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 480, in handle_argv
    return self.execute(command, argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celery.py", line 412, in execute
    ).run_from_argv(self.prog_name, argv[1:], command=argv[0])
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/worker.py", line 220, in run_from_argv
    self.maybe_detach([command] + argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/worker.py", line 227, in maybe_detach
    detached_celeryd(self.app).execute_from_commandline(argv)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celeryd_detach.py", line 96, in execute_from_commandline
    **vars(options)
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/bin/celeryd_detach.py", line 32, in detach
    after_forkers=False):
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/platforms.py", line 334, in open
    self.after_chdir()
  File "/Users/fortiz/pyenvs/notify/lib/python3.6/site-packages/celery/platforms.py", line 411, in after_chdir_do
    logfile and open(logfile, 'a').close()
FileNotFoundError: [Errno 2] No such file or directory: '/var/logs/worker2.log'
```
As you can see read at the beginning from the config file, but the workers print `-- 1` that's means the value is taken from the command line, but the error show a default value ``` FileNotFoundError: [Errno 2] No such file or directory: '/var/logs/worker2.log'```

Reply:
That is because this path `/var/logs` is set by default.
I've already told you that when the Celery start the CLI arguments is called again because the arguments to start it. That is why I had added the last commit in this PR.

Reply:
👍